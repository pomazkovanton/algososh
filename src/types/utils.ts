export type TMethod = "ascending" | "descending";

export type TQueue = {
  items: string[];
  head: number;
  tail: number;
};

export type TQueueStatus = "enqueue" | "dequeue" | "clear" | null;

export type TListStatus =
  | "add-in-head"
  | "add-in-tail"
  | "remove-from-head"
  | "remove-from-tail"
  | "add-by-index"
  | "remove-by-index"
  | null;
