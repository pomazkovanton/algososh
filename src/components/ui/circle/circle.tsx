import React from "react";
import styles from "./circle.module.css";
import { ElementStates } from "../../../types/element-states";

interface CircleProps {
  state?: ElementStates;
  letter?: string;
  head?: string | React.ReactElement | null;
  index?: number;
  tail?: string | React.ReactElement | null;
  tailType?: "string" | "element";
  extraClass?: string;
  isSmall?: boolean;
}

export const Circle: React.FC<CircleProps> = ({
  state = ElementStates.Default,
  letter,
  head,
  index,
  tail,
  extraClass = "",
  isSmall,
}) => {
  return (
    <div className={`${styles.content} ${extraClass}`}>
      <div
        className={`text text_type_input text_color_input mb-4 ${styles.absolute} ${styles.head} ${
          styles[typeof head === "string" ? "string" : "element"]
        }`}
        data-cy="circle-head"
      >
        {head}
      </div>
      <div className={`${styles.circle}  ${isSmall ? styles.small : ""} ${styles[state]}`} data-cy="circle-main">
        <p className={`text text_type_circle text_color_input ${styles.letter}`} data-cy="circle-text">{letter}</p>
      </div>
      <p className={`text text_type_input text_color_input mt-4 ${styles.absolute} ${styles.index}`} data-cy="circle-index">
        {index?.toString()}
      </p>
      <div
        className={`text text_type_input text_color_input mt-4 ${styles.absolute} ${
          index?.toString() ? styles.tail60 : styles.tail30
        } ${styles[typeof tail === "string" ? "string" : "element"]}`}
        data-cy="circle-tail"
      >
        {tail}
      </div>
    </div>
  );
};
