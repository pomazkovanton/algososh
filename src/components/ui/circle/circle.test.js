import renderer from 'react-test-renderer';

import { Circle } from './circle';
import { ElementStates } from '../../../types/element-states';

describe('Компонент <Circle />', () => {
    it('отрисовывается без буквы', () => {
        const tree = renderer
          .create(<Circle />)
          .toJSON();
          expect(tree).toMatchSnapshot();
    }); 
    it('отрисовывается c буквами', () => {
        const tree = renderer
          .create(<Circle letter='Test'/>)
          .toJSON();
          expect(tree).toMatchSnapshot();
    }); 
    it('отрисовывается c head', () => {
        const tree = renderer
          .create(<Circle head='Test'/>)
          .toJSON();
          expect(tree).toMatchSnapshot();
    }); 
    it('отрисовывается c react-элементом в head', () => {
        const tree = renderer
          .create(<Circle head={<Circle/>}/>)
          .toJSON();
          expect(tree).toMatchSnapshot();
    }); 
    it('отрисовывается c tail', () => {
        const tree = renderer
          .create(<Circle tail='Test'/>)
          .toJSON();
          expect(tree).toMatchSnapshot();
    }); 
    it('отрисовывается c react-элементом в tail', () => {
        const tree = renderer
          .create(<Circle tail={<Circle/>}/>)
          .toJSON();
          expect(tree).toMatchSnapshot();
    }); 
    it('отрисовывается c index', () => {
        const tree = renderer
          .create(<Circle index='0'/>)
          .toJSON();
          expect(tree).toMatchSnapshot();
    }); 
    it('отрисовывается с пропсом isSmall', () => {
        const tree = renderer
          .create(<Circle isSmall/>)
          .toJSON();
          expect(tree).toMatchSnapshot();
    }); 
    it('отрисовывается в состоянии default', () => {
        const tree = renderer
          .create(<Circle state={ElementStates.Default}/>)
          .toJSON();
          expect(tree).toMatchSnapshot();
    }); 
    it('отрисовывается в состоянии changing', () => {
        const tree = renderer
          .create(<Circle state={ElementStates.Changing}/>)
          .toJSON();
          expect(tree).toMatchSnapshot();
    }); 
    it('отрисовывается в состоянии modified', () => {
        const tree = renderer
          .create(<Circle state={ElementStates.Modified}/>)
          .toJSON();
          expect(tree).toMatchSnapshot();
    }); 
});