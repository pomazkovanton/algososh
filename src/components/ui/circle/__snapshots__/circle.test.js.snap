// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Компонент <Circle /> отрисовывается c head 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head string"
  >
    Test
  </div>
  <div
    className="circle   default"
    data-cy="circle-main"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Компонент <Circle /> отрисовывается c index 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
    data-cy="circle-main"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  >
    0
  </p>
  <div
    className="text text_type_input text_color_input mt-4 absolute tail60 element"
  />
</div>
`;

exports[`Компонент <Circle /> отрисовывается c react-элементом в head 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  >
    <div
      className="content "
    >
      <div
        className="text text_type_input text_color_input mb-4 absolute head element"
      />
      <div
        className="circle   default"
        data-cy="circle-main"
      >
        <p
          className="text text_type_circle text_color_input letter"
        />
      </div>
      <p
        className="text text_type_input text_color_input mt-4 absolute index"
      />
      <div
        className="text text_type_input text_color_input mt-4 absolute tail30 element"
      />
    </div>
  </div>
  <div
    className="circle   default"
    data-cy="circle-main"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Компонент <Circle /> отрисовывается c react-элементом в tail 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
    data-cy="circle-main"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  >
    <div
      className="content "
    >
      <div
        className="text text_type_input text_color_input mb-4 absolute head element"
      />
      <div
        className="circle   default"
        data-cy="circle-main"
      >
        <p
          className="text text_type_circle text_color_input letter"
        />
      </div>
      <p
        className="text text_type_input text_color_input mt-4 absolute index"
      />
      <div
        className="text text_type_input text_color_input mt-4 absolute tail30 element"
      />
    </div>
  </div>
</div>
`;

exports[`Компонент <Circle /> отрисовывается c tail 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
    data-cy="circle-main"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 string"
  >
    Test
  </div>
</div>
`;

exports[`Компонент <Circle /> отрисовывается c буквами 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
    data-cy="circle-main"
  >
    <p
      className="text text_type_circle text_color_input letter"
    >
      Test
    </p>
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Компонент <Circle /> отрисовывается без буквы 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
    data-cy="circle-main"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Компонент <Circle /> отрисовывается в состоянии changing 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   changing"
    data-cy="circle-main"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Компонент <Circle /> отрисовывается в состоянии default 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
    data-cy="circle-main"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Компонент <Circle /> отрисовывается в состоянии modified 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   modified"
    data-cy="circle-main"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Компонент <Circle /> отрисовывается с пропсом isSmall 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle  small default"
    data-cy="circle-main"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;
